mmcv/ops/csrc/common/pytorch_device_registry.hpp

```diff
@@ -0,0 +1,135 @@
+#ifndef PYTORCH_DEVICE_REGISTRY_H
+#define PYTORCH_DEVICE_REGISTRY_H
+
+#include <torch/extension.h>
+
+#include <cassert>
+#include <functional>
+#include <map>
+#include <type_traits>
+
+inline std::string GetDeviceStr(const at::Device& device) {
+  std::string str = DeviceTypeName(device.type(), true);
+  if (device.has_index()) {
+    str.push_back(':');
+    str.append(std::to_string(device.index()));
+  }
+  return str;
+}
+
+// Registry
+template <typename F, F f>
+class DeviceRegistry;
+
+template <typename Ret, typename... Args, Ret (*f)(Args...)>
+class DeviceRegistry<Ret (*)(Args...), f> {
+ public:
+  using FunctionType = Ret (*)(Args...);
+
+  void Register(at::DeviceType device, FunctionType function) {
+    entries_.insert({device, function});
+  }
+
+  FunctionType Find(at::DeviceType device) const {
+    auto it = entries_.find(device);
+    if (it != entries_.end()) {
+      return it->second;
+    }
+    return nullptr;
+  }
+
+  static DeviceRegistry& instance() {
+    static DeviceRegistry inst;
+    return inst;
+  }
+
+ private:
+  DeviceRegistry() = default;
+  std::map<at::DeviceType, FunctionType> entries_;
```

Comment:
Given that `at::DeviceType` is an int8_t enum that has range `[0, COMPILE_TIME_MAX_DEVICE_TYPES)`, may be we can replace the map with a null initialized constant sized array which would be much faster that `std::map`